Publish GitHub release from master branch

Gudahtt opened the pull request:
This ensures that changes made on `develop` since branching for the release are not included. It also ensures that the final release sourcemaps line-up correctly (they were always build on master)`.

Gudahtt committed:
Publish GitHub release from master branch

This ensures that changes made on `develop` since branching for the
release are not included. It also ensures that the final release
sourcemaps line-up correctly (they were always build on master)`.

kumavis reviewed (commented):
having trouble understanding the flow

seems like `job-publish-release` and `create_github_release` should be two parts of the same flow

Gudahtt commented:
Yes, they probably should be. Good point - that might be worth changing now, to ensure they aren't separated again.

Gudahtt committed:
Consolidate publish jobs

The jobs `job-publish-release` and `create_github_release` both handle
different parts of publishing a release. They have been consolidated
into a single `job-publish-release` job.

kumavis reviewed (dismissed):
great! still want to get an OK from @whymarrh first

as long as they are together, i dont mind as much what branch they build from

kumavis commented:
also relevant: im of the opinion we should get rid of the `master` branch (or rename `develop` to `master`). i think its all pain no gain. automated releases now correctly set a tag, so most of the reason to have master is covered.

kumavis commented:
so maybe for this PR, now that they are a single task, we can have it run on `develop` or the feature branch. Alternatively, if its hard to architect with circleCI, we could try github actions. I think that lets us hook into richer actions like PR merge.

Gudahtt commented:
The problem with targeting the release branch at `develop` and building upon merge is that we'd then include anything merged into `develop` since the release branch was created.

I'm not sure how we could trigger either CircleCI or GitHub Actions to build on the last commit of the branch being merged, _before_ the merge. If we could find a way to do that, that'd be great.

Instead we might need to come up with an alternate means of triggering the release - some event other than the merging of the release branch. What I was considering using is the `git tag` itself - we could build upon that being added, and even have CircleCI/Actions merge the release branch for us. So then the workflow would be to add the tag when the release was ready, and the rest would be automated. That's a bit of a weird workflow though - there are alternatives we could use that involve reviews/comments being the trigger instead, that we'd need either a GitHub App/bot or GitHub Actions to facilitate.

whymarrh reviewed (changes requested):
We'll need to do something with the grep pattern we're using in the script:

https://github.com/MetaMask/metamask-extension/blob/db3d7542058df774f4dd772548f5267fd23e1524/.circleci/scripts/release-create-gh-release#L31

Gudahtt committed:
Update release script to expect a merge commit

The release script was originally written to be run on `develop`, so it
expected the current commit to be a result of `Squash & Merge`. Now
that it's run on `master`, it will generally be run against a merge
commit.

The version is now extracted from the commit message using a regular
expression that should work on all version of Bash v3+, and should be
tolerant of both merge commits and `Squash & Merge` commits.

Gudahtt committed:
Target `master` with release PR

`master` is now targetted by the release PR instead of `develop`, as
the release has to be created from the master branch.

The update to `develop` is handled after the release by a PR from
`master` to `develop`, which is created automatically after the
release.

whymarrh commented on .circleci/scripts/release-create-gh-release:
```diff
@@ -28,24 +28,24 @@ function install_github_cli ()
             --attach metamask-chrome-*.zip \
             --attach metamask-firefox-*.zip \
-            --message "${commit_words[0]} ${commit_words[1]#v}" \
+            --message "Version ${tag##v}" \
```
Small nitpick but we don't need to strip the `v` prefix here:

```suggestion
            --message "Version ${tag}" \
```

The 1st (0-indexed) match will be without the `v` as per the regex above.

Gudahtt replied:
Oops, I had intended to capture the `v` as well. I'll fix the regex instead, otherwise line 46 is incorrect.

Gudahtt committed:
Include `v` in regex capture group

whymarrh reviewed (approved):
LGTM
